src/RequestCookies/RequestCookiesProvider.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Cookies\RequestCookies;
+
+use Yiisoft\Cookies\RequestCookies\Exception\RequestCookieCollectionNotSetException;
+
+/**
+ * Stores request for further consumption by attribute handlers.
+ */
+final class RequestCookiesProvider implements RequestCookiesProviderInterface
+{
+    /**
+     * @var RequestCookies|null The collection.
+     */
+    private ?RequestCookies $cookieCollection = null;
+
+    public function set(RequestCookies $cookieCollection): void
```

Comment:
I'd have an empty collection in this class and merge two collections instead. 
Moreover it can be used in two ways: 
The first is as you did
The second is the similar that error-handler has: set parameters in the app and add them back to request in a post-middleware later. Wdyt?